packages/plugin-core/src/commands/MoveHeader.ts

```diff
@@ -0,0 +1,311 @@
+
+type CommandInput =
+  | {
+      dest?: NoteProps;
+    }
+  | undefined;
+type CommandOpts = {
+  origin: NoteProps;
+  nodesToMove: Node[];
+  modifiedOriginTree: Node;
+  originProc: Processor;
+  engine: EngineAPIService;
+} & CommandInput;
+type CommandOutput = {
+  updated: NoteProps[];
+} & CommandOpts;
+
+export class MoveHeaderCommand extends BasicCommand<
+  CommandOpts,
+  CommandOutput
+> {
+  key = DENDRON_COMMANDS.MOVE_HEADER.key;
+
+  headerNotSelectedError = new DendronError({
+    message: "You must first select the header you want to move.",
+  });
+
+  noActiveNoteError = new DendronError({
+    message: "No note open.",
+  });
+
+  getProc = (engine: EngineAPIService, note: NoteProps) => {
+    return MDUtilsV5.procRemarkParse(
+      { mode: ProcMode.FULL },
+      {
+        engine,
+        fname: note.fname,
+        vault: note.vault,
+        dest: DendronASTDest.MD_DENDRON,
+      }
+    );
+  };
+
+  /**
+   * Recursively check if two given node is identical.
+   * At each level _position_ is omitted as this can change if
+   * you are comparing from two different trees.
+   * @param a first {@link Node} to compare
+   * @param b second {@link Node} to compare
+   * @returns boolean
+   */
+  hasIdenticalChildren = (a: Node, b: Node): boolean => {
+    if (_.isEqual(Object.keys(a).sort(), Object.keys(b).sort())) {
+      const aOmit = _.omit(a, ["position", "children"]);
+      const bOmit = _.omit(b, ["position", "children"]);
+      if (_.isEqual(aOmit, bOmit)) {
+        if (_.has(a, "children")) {
+          return _.every(a.children as Node[], (aChild, aIndex) => {
+            const bChild = (b.children as Node[])[aIndex];
+            return this.hasIdenticalChildren(aChild, bChild);
+          });
+        }
+        return true;
+      } else {
+        return false;
+      }
+    } else {
+      return false;
+    }
+  };
+
+  async gatherInputs(opts: CommandInput): Promise<CommandOpts | undefined> {
+    const engine = getEngine();
+    const { editor, selection } = VSCodeUtils.getSelection();
+    if (!editor) throw this.noActiveNoteError;
+    if (!selection) throw this.headerNotSelectedError;
+
+    // const text = editor.document.getText();
+    const line = editor.document.lineAt(selection.start.line).text;
+    const maybeNote = VSCodeUtils.getNoteFromDocument(editor.document);
+    if (!maybeNote) {
+      throw this.noActiveNoteError;
+    }
+
+    const proc = this.getProc(engine, maybeNote);
+    const parsedLine = proc.parse(line);
+    let targetHeader: Heading | undefined;
+    visit(parsedLine, [DendronASTTypes.HEADING], (heading: Heading) => {
+      targetHeader = heading;
+      return false;
+    });
+    if (!targetHeader) {
+      throw this.headerNotSelectedError;
+    }
+    const tree = proc.parse(maybeNote.body);
+    let headerFound = false;
+    let foundHeaderIndex: number | undefined;
+    let nextHeaderIndex: number | undefined;
+    visit(tree, (node, index) => {
```

Comment:
- can we abstract this away in [RemarkUtils](https://github.com/dendronhq/dendron/blob/e83f4945c0838221bb635ca69ce9b905c9d3f6d0/packages/engine-server/src/markdown/remark/utils.ts#L822-L822)?

```ts
extractHeaderBlock(targetHeader: Heading) => {blockStartIndex: number, blockEndIndex: number }
```